### PR TITLE
feat: move gun into worker

### LINKED ISSUE #1886: [Demand] On-demand open Gun connection and run it in Web Worker

### PR DESCRIPTION
**Pull request by Jack-Works**
close #1886 cc @septs 

(not tested yet)

### CODE REVIEW

File: packages/maskbook/src/extension/background-script/CryptoServices/encryptTo.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as Alpha38 from '../../../crypto/crypto-alpha-38'
-import * as Gun2 from '../../../network/gun/version.2'
+import { GunAPI as Gun2 } from '../../../network/gun/'
```
Comment (septs) [inline packages/maskbook/src/extension/background-script/CryptoServices/encryptTo.ts:2]:
```suggestion
import { GunAPI as Gun2 } from '../../../network/gun'
```

**Review by septs — DISMISSED** [review]

**Comment by septs** [comment]
waiting for me to test.

File: packages/maskbook/src/web-workers/OnDemandWorker.ts [hunk 2]
```diff
@@ -27,31 +25,61 @@ export class OnDemandWorker extends Worker {
         }, Math.min(this.inactiveTimeToTerminate, WorkerCheckTerminateInterval))
     }
     protected log(...args: any[]) {
-        // console.log(`OnDemandWorker ${this.id}`, ...args)
+        console.log(`OnDemandWorker ${this.id}`, ...args)
```
Comment (septs) [inline packages/maskbook/src/web-workers/OnDemandWorker.ts:28]: provide a worker name field in log?

**Comment by Jack-Works** [comment]
There're still two problems to be resolved before merging:

- [ ] Implement constructible `EventTarget` for iOS 14-
- [ ] Figure out how to config gun to use indexed db

**Comment by Jack-Works** [comment]
Another question, how often should we check if a worker should be terminated? (`WorkerCheckTerminateInterval` in the code) and how long should a worker live before it's being terminated `InactiveTimeToTerminateDefault` in the code?

File: packages/maskbook/src/web-workers/OnDemandWorker.ts [hunk 3]
```diff
@@ -27,31 +27,77 @@ export class OnDemandWorker extends Worker {
         }, Math.min(this.inactiveTimeToTerminate, WorkerCheckTerminateInterval))
     }
     protected log(...args: any[]) {
-        // console.log(`OnDemandWorker ${this.id}`, ...args)
+        console.log(`OnDemandWorker ${this.init[1]?.name}`, ...args)
```
Comment (septs) [inline packages/maskbook/src/web-workers/OnDemandWorker.ts:30]: name + unique id

Comment (Jack-Works) [reply packages/maskbook/src/web-workers/OnDemandWorker.ts:30]: It's callsite duty. This options is inherit from the native Worker.

**Comment by Jack-Works** [comment]
I have one last small problem before merge. Need help due to my shortage this week @septs . I still don't know how to enable indexed db integration for gun meanwhile have the ability to connect to the peer.

**Comment by Jack-Works** [comment]
ok everything is ready. plz review this @septs @guanbinrui @zhouhanseng . please use "squash and merge" instead of "rebase and merge" for this PR

**Review by septs — APPROVED** [review]

**Comment by septs** [comment]
image

CC @yisiliu @asonmine

**Comment by Tedko** [comment]
Is the traffic normal..?

**Comment by septs** [comment]
@Tedko current 35 peers online. (normal)

see https://gun-jp.maskbook.com/stats.html

**Comment by Jack-Works** [comment]
Does it mean this PR is very successful in saving trafics?

**Pull request merged**